### ISSUE TITLE
terminal stays open after process exit

### ISSUE THREAD

**Issue by payonel** [Microsoft/vscode#63813]
Initially I thought this was an issue with MIEngine, but they report that it is up to VSCode to properly close the external console after debugging

https://github.com/Microsoft/MIEngine/issues/807

Version: 1.29.1
Commit: bc24f98b5f70467bc689abf41cc5550ca637088e
Date: 2018-11-15T19:07:43.495Z
Electron: 2.0.12
Chrome: 61.0.3163.100
Node.js: 8.9.3
V8: 6.1.534.41
Architecture: x64

repro
1. debug cpp with external terminal
2. when the program ends, the terminal remains open

"Press any key to continue.." is holding these zombie terminals open (we get a new one with each run/debug run)

**Comment by killerkalamari** [comment]
I came to report the same problem on the Linux version. This new behavior seemed to coincide with a fix for the Linux terminal emulator selection (before it would always used xterm, regardless of settings... now it respects the setting, defaulting to x-terminal-emulator, but this "Press any key to continue..." thing came along for the ride). 

I'm assuming that it was always supposed to be doing this "Press any key" thing in the first place, and that this was a bug that they "fixed" along with fixing the terminal selection problem. However, I wish there were a way for me to disable it, because I end up with numerous lingering terminals during intense debugging sessions. Two of my co-workers also came to me annoyed with it and asking me to see if I can hack around it somehow.

**Comment by guoyiteng** [comment]
I would like to add that this problem also exists on the internal console. Each debugging session will launch a new terminal and when the program ends, the internal terminal stays open with text
```
[1] + Done                       "/usr/bin/gdb" --interpreter=mi --tty=${DbgTerm} 0<"/tmp/Microsoft-MIEngine-In-rxo7l3v7.xwx" 1>"/tmp/Microsoft-MIEngine-Out-c3yh2g6c.61r"
```
After debugging the program for a while, there would be a bunch of internal terminals. I have to manually close all of them, which is pretty annoying. Sadly, the workaround @killerkalamari mentioned does not work for the internal terminal.

This problem only happens on Linux. The internal terminal on macOS could exit after the program ends.

**Comment by pingala98** [comment]
@guoyiteng  please let me know if you find any workaround, for someone who does competitive programming on vs code like me, it is VERY MUCH ANNOYING to close 2 terminal tabs after each debug session.

**Comment by therocode** [comment]
Same here, after debugging for a while, I have 20+ terminals and VS Code even gives an error at 25 terminals because it's not allowed to allocate any more (on Windows 10 here). Would be good if running the program again reused the terminal

**Comment by hammypants** [comment]
+1 from me, would be nice to not have to manually kill terminals

**Comment by connor4312** [comment]
@weinand I notice we 'intentionally' add `&& wait` and the like to terminals. Should we remove that wholesale, or or make it a user option? I.e. is this bug actually a feature request for a setting, or a change in default behavior?

**Comment by killerkalamari** [comment]
I hope it can be a setting, because I end up turning off my workaround sometimes when I've really messed things up and need to see why it crashed.

**Comment by connor4312** [comment]
Maybe a better approach would be to kill any existing integrated terminal (for the same configuration?) when launching a new session, so that you would just have 1 extra terminal hanging around, not 100.

**Comment by weinand** [comment]
- just a clarification: this issue is about the **external** terminal, not the integrated terminal
- driving external terminals from VS Code on different platforms is challenging. For instance we couldn't find a way on all platforms to reuse an external terminal for multiple sessions because most APIs only support to pass a single shell command on terminal creation. So feeding additional commands into an existing terminal is difficult. Exception: the "Terminal" application on macOS provides AppleScript API for determining whether a terminal is "busy"  and for feeding additional input.
- initially we were closing the terminal immediately at the end of a debug session but with that behavior user complained that they couldn't see runtime errors or program output. That's the reason why the `"Press any key to continue..." && wait` was added. IIRC the idea was to keep the terminal open if there are issues (exit code != 0) and to close it otherwise.

---

Yes, killing the "previous" terminal when creating a "new" terminal sounds like a good approach to avoid accumulating lots of terminals. But we need to be careful not to kill terminals that are still in use. Using the "configuration name" as a way to find terminal candidates that can be closed only works, if we continue to prevent that launch configs are launched only once (we are planning to lift this - see #110226).

**Comment by vdedios** [comment]
Why not adding an extra option in `launch.json` and let users decide whether they want or not to have terminal debugging closed when debug ends?

Something like `closeTermOnEnd: true/false`

**Comment by weinand** [comment]
Support for external terminals is provided by individual debug extensions. VS Code doesn't know about them and cannot provide a generic property in the launch.json.

**Comment by Bernd-L** [comment]
I'd suggest introducing a new property in "launch.json" via `BaseTaskConfiguration`:  

```TypeScript
interface BaseTaskConfiguration {
  // ...
  
  /**
  * If an external terminal should wait after
  * its command has exited or not.
  */
  waitAfterExternExit: boolean
}
```
with the default value `true` or something like that.

**Comment by weinand** [comment]
@Bernd-L please see my previous comment. I don't understand how your "BaseTaskConfiguration" would address the layering issue?

**Comment by Bernd-L** [comment]
> @Bernd-L please see my previous comment. I don't understand how your "BaseTaskConfiguration" would address the layering issue?

I meant that extending the existing one at <https://github.com/microsoft/vscode-docs/blob/master/docs/editor/tasks-appendix.md#schema-for-tasksjson> might be a good idea, to include the new property `waitAfterExternExit`.

**Comment by connor4312** [comment]
> Yes, killing the "previous" terminal when creating a "new" terminal sounds like a good approach to avoid accumulating lots of terminals. But we need to be careful not to kill terminals that are still in use. Using the "configuration name" as a way to find terminal candidates that can be closed only works

The challenge I face here is that there is nothing that prevents the debugger from executing a `runInTerminal` request that executes a command that outlives the debug session. If we can assume this is the case it also gives me a nice solution to https://github.com/microsoft/vscode/issues/112055, but I didn't take that approach initially because I am not sure if there are any consumers that do this.

**Comment by vdedios** [comment]
Just for now, I came up with kind of a workaround that solves accumulating tons of opened terms after a few debugging sessions. You just can use VSCode's `postDebugTask` to launch a task when debug session ends. The task could look something like:

in `tasks.json`:
```
        (...)
        {
            "label": "Close term",
            "type": "shell",
            "command": "killall -SIGTERM Terminal"
        }
```

then, in `launch.json` add `"postDebugTask": "Close term"`

If you are working with Terminal as well as debugging you could play around with pids and adding a shell command to kill only desired terms.

There is another issue in `MacOS` though, when you kill all terms and reopen them, last session is restored so you end up with more and more terms opened each time you restart a new session. To avoid it, I recommend following this [solution on stack exchange](https://apple.stackexchange.com/questions/187765/clear-terminal-and-prevent-restoration). Enjoy!

**Comment by zeroxia** [comment]
I like the idea of closing previous terminal when a new debugging terminal is launched.
It should be easy to have a handle to last debugging terminal, and close it when launching a new one.
This means there is no cluttering of a lot of obsolete debugging terminals, and user can always check the terminal state after debugging done.
Please make an option in `launch.json` to do so! Thanks!

**Comment by pm100** [comment]
@weinand You explicitly state that the 'not automatically closing' behavior was added 

> That's the reason why the "Press any key to continue..." && wait was added.

All we are asking for is for a switch somewhere to make that optional, because it can be a real pain.

**Comment by AlexandreRouma** [comment]
Just adding to what's been said above, accumulating terminals after multiple debug run is extremely annoying.
VSCode really needs to add a switch to disable that anti-feature...
